README
These are the scripts to support using a caching proxy to build OpenWRT,
so you can have reproducable, auditable builds.  Note that this does not
by itself make the code secure - it just makes it *theoretically possible*
for you to perform a security audit on it.

It also saves bandwidth and time - with the mirror, you don't have to wait
on 300-odd megs of source code per architecture built (not counting the
svn checkout of openwrt itself, too) each time you do a clean build.

With the caching proxy, if you tweak something and then rebuild the firmware, 
you're building from the same sources you used last time.  Note that this is
a mixed blessing - your build isn't going to be broken by bugs introduced
by upstream developers, but you're also not going to get upstream bugfixes
unless you explicitly make an effort to do so.

Assumptions:

User 'sudomesh' builds the firmware.  User 'sudomirror' maintains the mirror.
A third user has root.

There is an http server on localhost, and ~/public_html/ maps to ~username
URLs.

~/bin/ is in the default path ahead of systemwide binaries.

You have svnrdump.  (If your systemwide version of subversion does not include
svnrdump, you can compile a newer version and drop the svnrdump binary in
~sudomirror/bin/ )

To use (from the build proxy directory):

As root, create the users 'sudomesh' and 'sudomirror'.
        adduser sudomesh
        adduser sudomirror
And set up the firewall rules:
        ./firewall.sh

As sudomesh:
        mkdir ~/bin
        ln -sf sudomesh-bin/* ~/bin/

As sudomirror, configure the mirror:
        mkdir ~/bin
        ln -sf sudomirror-bin/* ~/bin/
        prep-cloneserver.sh

If prep-cloneserver.sh returns 'command not found', you probably need to log
out and log back in to both the sudomesh and sudomirror accounts.  If it STILL
returns 'command not found' after logging back in, then you'll need to put
~/bin/ in your path in both the sudomesh and sudomirror accounts.

As sudomirror, start the mirror:
        build-proxy.py
        cloneserver.py --allow-write

Only use --allow-write when you actively intend to permit the mirror to be
updated.  (Necessary on the first build, NOT necessary on subsequent builds.)

As sudomesh, check out the OpenWRT firmware:
        git clone https://github.com/sudomesh/sudowrt-firmware.git

And build it:
        cd sudowrt-firmware
        ./prepare

The first time, it will download all the sources and cache them.  If you
want to start the cloneserver without allowing it to download new files,
run it without --allow-write.

--- ---

Other caveats:

This runs as two separate users to prevent an attacker from concealing
modifications to the sources from you.  An attacker who controls the right
parts of the network and can create an MD5 collision, or who otherwise
controls the openwrt firmware source, can still execute whatever code
they like as sudomesh.

prepare and the OpenWRT build scripts try multiple mirrors when looking
for sources to download.  The first source tried is not always valid; an
attacker who controls the right parts of the net could, when --allow-write
is set on the cloneserver, inject their own code in the place of higher
priority mirrors.

The scripts also assume that git, svn, svnsync, svnrdump, and svnadmin
never execute code downloaded from the remote.  I do not know if this is
true, I'm just assuming that it is!



Changelog.txt
[ 0.2.1+ ]
* HTTP repo cloning warning now only displayed if actually trying to clone
an HTTP repo.
* cloneserver now closes connections after sending the ok/fail message.
* git stub now tries to get the mirror to clone things.
* Cloneserver now logs attempts.
* Wget mirrorer now accepts / in URLs
* prep-cloneserver script added.
* cloneserver: wget mirroring works.
* cloneserver: wget mirroring now uses preexisting files if one exists.
* Build script: wget mirroring works.
* Improved logging.
[ 0.2.3 ]
* svn mirroring now distinguishes between http and svn protocol mirrors.
* svn mirroring now works correctly when mirroring something that's a path
under the full repository - e.g. http://svn.nomi.cz/svn/isteve/hotplug2 is
actually in the http://svn.nomi.cz/svn/isteve/ repo.  Symlinks from
(mirror root)/svn-(protocol)-roots/(domain)/(url path) to the actual mirror
directory are created when mirroring completes.
* A number of errors and warnings now say which app they're from.
[ 0.2.4 ]
* _ accepted in svn clone urls
* Cloneserver scripts now set umask 022.
* Now uses svnrdump for Subversion mirroring, which seems to be much much
faster.
[ 0.2.5 ]
* hackish fixup for OpenWRT subversion sync broken by fast mirror, so removed.
* svnrdump doesn't work with HTTP svn mirrors, so use svnsync when cloning
non-svn:// svn URLs.
* git now mirrors to per-protocol paths.
* Can now clone git https repos.
[ 0.3 ]
* Documentation clarifications.
[ 0.3.1 ]


firewall.sh
#!/bin/sh

# Flush the output table; note that this will be bad if anything else is
# sharing the server.
iptables -t nat -F OUTPUT

# All DNS queries go to the local machine.  That way DNS will work later...
iptables -t nat -m owner -A OUTPUT --uid-owner sudomesh -p tcp --dport 53 -j DNAT --to-destination 127.0.0.1
iptables -t nat -m owner -A OUTPUT --uid-owner sudomesh -p udp --dport 53 -j DNAT --to-destination 127.0.0.1

# Redirect all attempts to access remote HTTP servers to localhost:8081.
iptables -t nat -m owner -A OUTPUT --uid-owner sudomesh -p tcp --dport 80 \! -d 127.0.0.1 -j DNAT --to-destination 127.0.0.1:8081
# Redirect all attempts to access remote HTTPS servers to localhost:443.
iptables -t nat -m owner -A OUTPUT --uid-owner sudomesh -p tcp --dport 443 \! -d 127.0.0.1 -j DNAT --to-destination 127.0.0.1:443

# And reject everything else the build UID sends beyond localhost.
iptables -m owner -A OUTPUT --uid-owner sudomesh \! -d 127.0.0.1 -j REJECT


mirror-bin/build-proxy.py
#!/usr/bin/env python

"""
To use:
Step 0: Configure your archive URL correctly.
"""
ARCHIVE = "http://localhost/~sudomirror/mirror/http/"
"""
The docs will assume that the firmware is built by user sudomesh and
the mirror is owned by user sudomirror.

redirect all network access from the build account to localhost.
    (See firewall.sh for an example.)

Put the git and svn rearrangers in ~sudomesh/bin/ and make sure that
~sudomesh/bin/ is at the beginning of sudomesh's path.  You may need to log
out and back in, or rehash.

Start (as the mirror user) build-proxy.py.

"""

import sys
import select
import socket

sock = socket.socket()
sock.setsockopt(socket.SOL_SOCKET,socket.SO_REUSEADDR,1)
sock.bind(("127.0.0.1",8081))
sock.listen(2)

def read_line(sock):
    ret = ''
    while 1:
        c = sock.recv(1)
        ret += c
        if c=='\n':
            return ret
        elif not c:
            return ret

def handle_req(sock):
    while 1:
        line = read_line(sock)
        if line.endswith("\012"):
            line = line[:-1]
        if line.endswith("\015"):
            line = line[:-1]
        if line.startswith("GET ") and line.endswith(" HTTP/1.1"):
            URL = line[4:-9]
            print "URL:",URL
        elif line.startswith("Host: "):
            HOST = line[6:]
            print "Host:",HOST
        elif not line:
            sock.send("HTTP/1.1 301 Moved Permanently\015\012Location: "+ARCHIVE+HOST+URL+"\015\012\015\012")
            sock.shutdown(1)
            sock.close()
            return
        else:
            pass
            #print "Ignoring HTTP header",line

peer = sock.accept()
while peer:
    print "Connected",peer[1]
    URL = 'nil'
    HOST = 'nil'
    handle_req(peer[0])
    peer = sock.accept()


mirror-bin/git-clonerepo.pl
#!/usr/bin/perl

$MIR='/home/sudomirror/public_html/mirror';

die unless $#ARGV==0;
if ($ARGV[0]=~/^(git|http|https):\/\/(.*?)\/(.*)$/s) {
        $proto = $1;
        $domain = $2;
        $path = $3;
        if ($proto ne 'git') {
                $proto_dir = 'git-'.$proto;
        } else {
                $proto_dir = 'git';
        }
        if ($path=~/^(.*?);(.*)$/s) {
                $path = $1;
                $branch = $2;
        } else {
                $branch = 'master';
        }
        system('mkdir','-p',"$MIR/$proto_dir/$domain/$path");
        chdir ("$MIR/$proto_dir/$domain/$path") or die "$!\n";
        system('git', '--bare', 'init');
        print "[git-clonerepo.pl:] About to fetch.";
        system('git', '--bare', 'fetch', "$proto://$domain/$path", "$branch:$branch");
        system('git','update-server-info');
}



mirror-bin/svn-clonerepo.pl
#!/usr/bin/perl -T
$MIRROR_ROOT = "/home/sudomirror/public_html/mirror/";
chdir($MIRROR_ROOT) or die "chdir(): $!\n";
$ENV{PATH} = '/home/sudomirror/bin:/usr/local/bin:/usr/bin:/bin';
# cd ~/public_html/mirror/svn/ (Create it if needed)
$FAST = 1;

sub get_symlink_path($$$)
{
    my ($proto, $domain, $url_path) = @_;
    my $symlink_path = $MIRROR_ROOT."svn-$proto-roots/$domain";
    my @path = split('/',$url_path);
    for my $arg (@path) {
        $symlink_path .= '/'.$arg;
    }
    return $symlink_path;
}

sub do_clone($)
{
        my $proto;
        my $domain;
        my $url_path;
        if ($_[0]=~/^(svn|http):\/\/([a-zA-Z.]+)\/([\/0-9_a-z]+)$/s) {
                $proto = $1;
                $domain = $2;
                $url_path = $3;
        } else {
                die "[svn-clonerepo.pl] Cannot parse repository URL: $_[0]\n";
        }
        my $symlink_path = &get_symlink_path($proto,$domain,$url_path);
        if (-l $symlink_path) {
            print "[svn-clonerepo.pl:] $symlink_path already exists, skipping mirror.\n";
            return;
        }

        my $INFO = `svn info $proto://$domain/$url_path`;
        my $INFO_root = undef;
        my $INFO_url = undef;
        if ($INFO=~/^Repository Root: (.*?)$/m) {
            $INFO_root = $1;
        }
        if ($INFO=~/^URL: (.*?)$/m) {
            $INFO_url = $1;
        }
        die "No URL found in svn info\n" unless defined $INFO_url;
        die "No root found in svn info\n" unless defined $INFO_root;
        my $repo_path;
        if (substr($_[0],0,length($INFO_root)) eq $INFO_root) {
            $repo_path = $INFO_root;
        } else {
            die "Error: Cloning a repo that's not inside its own root?\n";
        }
        $repo_path=~s/^([a-z]+):\/\/([^\/]+)\///s;

        my @path = split('/',$repo_path);
        my $dir = $path[0];
        my $localdir = $MIRROR_ROOT.'svn';
        $localdir .= '-'.$proto if ($proto ne 'svn');

        mkdir("$localdir",0777);
        $localdir .= "/$domain";
        mkdir("$localdir",0777);
        for my $arg (@path) {
                $localdir .= '/'.$arg;
                mkdir("$localdir",0777);
        }
        print "[svn-clonerepo:] Debug: localdir $localdir\n";
        die "Could not create $localdir: $!\n" unless -d "$localdir";
        system('svnadmin', 'create', "$localdir");

        if ($FAST and $proto eq 'svn') {
                my $rev;
                if ($INFO=~/^Revision: ([0-9]+)$/m) {
                        $rev = $1;
                } else {
                        die "Error: Could not find revision in [$INFO].\n";
                }
                system("svnrdump dump -r$rev $proto://$domain/$repo_path >$MIRROR_ROOT/tmp/$$.svn.dump");
                if ($?) {
                        die "[svn-clonerepo.pl]: 'svnrdump dump -r$rev $proto://$domain/$repo_path >$MIRROR_ROOT/tmp/$$.svn.dump' failed with $?\n";
                }
                system("svnadmin load $localdir < $MIRROR_ROOT/tmp/$$.svn.dump");
                die "svnadmin load failed with $?\n" if $?;
                unlink("$MIRROR_ROOT/tmp/$$.svn.dump") or die "Could not delete tempfile.\n";
        } else {
                local *F;
                my $fn = "$localdir/hooks/pre-revprop-change";
                open(F,">$fn") or die "Cannot open $fn: $!\n";
                die "$!\n" unless print F "#!/bin/sh\n";
                die "$!\n" unless close(F);
                chmod(0755,$fn) or die "Error $! chmodding $fn\n";
                print "About to run 'svnsync init file://$localdir $proto://$domain/$repo_path\'\n";
                system("svnsync","init", "file://$localdir", "$proto://$domain/$repo_path");
                print "[svn-clonerepo:] Debug: svnsync init returned $?\n";
                system("svnsync", "sync", "file://$localdir");
                print "[svn-clonerepo:] Debug: svnsync sync returned $?\n";
        }
        if ($INFO_root ne $_[0]) {
            my @path = split('/',$symlink_path);
            pop(@path);
            my $dir = '';
            for my $arg (@path) {
                $dir .= "/$arg";
                mkdir($dir,0777);
            }
            if (!-d $dir) {
                die "[svn-clonerepo] $dir is not a dir.\n";
            }
            print "Debug: would like to symlink $localdir to $symlink_path";
            symlink($localdir,$symlink_path) or die "[svn-clonerepo:] symlink error: $!\n";
        }
}


for my $arg (@ARGV) {
        if ($arg eq '--slow') {
                $FAST = 0;
        } else {
                &do_clone($arg);
        }
}


mirror-bin/cloneserver.py
#!/usr/bin/env python

import sys
import select
import socket
import os
import re

ALLOW_UPDATE = ALLOW_WRITE = False
for arg in sys.argv[1:]:
    if arg=='--allow-write':
        ALLOW_WRITE = True
    elif arg=='--allow-update':
        ALLOW_UPDATE = True
    else:
        print "Unexpected argument",arg
        print "syntax: cloneserver.py [--allow-write] [--allow-update]"
        raise SystemExit(1)

if os.umask(022)!=022:
    print "Warning: umask was something other than 022.  Changed, so the cloned files will be readable."

MIRROR_ROOT='/home/sudomirror/public_html/mirror/'

sock = socket.socket()
sock.setsockopt(socket.SOL_SOCKET,socket.SO_REUSEADDR,1)
sock.bind(("127.0.0.1",8082))
sock.listen(2)

def read_line(sock):
    ret = ''
    while 1:
        c = sock.recv(1)
        ret += c
        if c=='\n':
            return ret
        elif not c:
            return ret

def clean_path(url):
    segs = url.split('/')
    ret = []
    for seg in segs:
        if '..' in seg:
           return None
        if seg.startswith('.'):
           return None
        if not seg:
           return None
        if '\012' in seg:
           return None
        if re.match('^([-A-Z_a-z0-9.]+)$',seg):
           if seg.startswith('-'):
               return None
           ret.append(seg)
        else:
           return None
    if ret:
        return '/'.join(ret)

def log(proto, result, data):
    """
    Each log line consists of (protocol) (result) (data), separated by spaces.
    There may be spaces in the log data.
    
    Allowed results:
    cloned - successfully created the initial clone of (url in data)
    in-cache - We already have a clone, using it.
    ok - clone successful or already existed.
    trying - Client requested clone of (url in data).
    updated - Updated a repository (git or svn)

    Error:(details) - An error occurred.
    """
    if '\012' in data:
        raise Exception("Newlines in log data not allowed.")
    f = open(os.path.join(MIRROR_ROOT,'mirror.log'),'ab')
    f.write(' '.join((proto,result,data))+"\n")
    f.close()

def logged(proto, func, url):
    log(proto,'trying',url)
    ret = func(url)
    if ret:
        log(proto,'sent',url)
    else:
        log(proto,'Error:Fail',url)
    return ret

def wget_clone(url):
    if url.startswith('http://'):
        path = clean_path(url[7:])
        if not path:
            raise Exception("Could not sanitize URL")
        dest_path = os.path.join(MIRROR_ROOT,'http',path)
        try:
            file(dest_path,'rb').close()
            log('wget','in-cache',url)
            return True
        except:
            pass
        if not ALLOW_WRITE:
            print "Error: Cannot mirror."
            return False
        spath = dest_path.rsplit('/',1)
        try:
            os.makedirs(spath[0])
        except:
            os.listdir(spath[0])
    else:
        log('wget','Error:Protocol',url)
        raise Exception("Unexpected URL protocol")
    try:
        os.makedirs(os.path.join(MIRROR_ROOT,'tmp'))
    except:
        pass
    tmpfilename = os.path.join(MIRROR_ROOT,'tmp/wget.'+str(os.getpid())+'.tmp')
    if os.spawnvp(os.P_WAIT,'wget',['wget',url,'-O',tmpfilename]):
        print "An error occurred downloading",url
        try:
            os.unlink(tmpfilename)
        except Exception,e:
            print "Delete of",tmpfilename,"failed:",e
            pass
        log('wget','Error:download-failed',url)
        return False
    os.rename(tmpfilename,dest_path)
    return True

def svn_clone(url):
    proto, path = url.split('://',1)
    if proto not in ('http','svn'):
        print "Bad protocol in",url
        raise Exception("Unexpected protocol.\n")
    if proto == 'svn':
        protodir = 'svn'
    else:
        protodir = 'svn-'+proto
    path = clean_path(path)
    if not path:
        print "Could not clean path in",url
        raise Exception("Path cleaning error.")
    dest_path = os.path.join(MIRROR_ROOT,protodir,path)
    try:
        os.listdir(dest_path)
        exists = True
    except Exception,e:
        print "cloneserver.py: can't read",dest_path,'-',e
        exists = False
    if not exists:
        if ALLOW_WRITE:
            print "About to clone"
            retv = os.spawnvp(os.P_WAIT,'svn-clonerepo.pl',['svn-clonerepo.pl',url])
            if retv:
                print "FAIL",url,retv
                return False
            log('svn','cloned',url)
            return True
        print "Write disallowed, not mirroring."
        return False
    elif ALLOW_UPDATE:
        print "[cloneserver.py:] Subversion sync."
        retv = os.spawn(os.P_WAIT, 'svnsync', ['svnsync','sync','file://'+MIRROR_ROOT+protodir+'/'+path])
        if retv:
            print "[cloneserver.py:] svnsync of",url,'failed with return value',retv
            return False
        log('svn','updated',url)
        return True
    else:
        print "Cached."
        log('svn','in-cache',url)
        return True

def git_clone(url):
    proto, path = url.split('://',1)
    if proto not in ('git','http','https'):
        print "[cloneserver.py:] Bad protocol in",url
        raise Exception("Unexpected protocol for git clone.")
    path = clean_path(path)
    if not path:
        print "Could not clean path in",url
        raise Exception("Could not clean path.")
    if proto != 'git':
        proto_dir = 'git-'+proto
    else:
        proto_dir = proto
    dest_path = os.path.join(MIRROR_ROOT,proto_dir,path)
    try:
        os.listdir(dest_path)
        exists = True
    except:
        exists = False
    if not exists:
        if ALLOW_WRITE:
            print "About to create initial mirror of git repo."
            retv = os.spawnvp(os.P_WAIT,'git-clonerepo.pl',['git-clonerepo.pl',url])
            if retv:
                print "FAIL",url,retv
                return False
            log('git','cloned',url)
            return True
        print "Write disallowed, not mirroring."
        return False
    elif ALLOW_UPDATE:
        retv = os.spawnvp(os.P_WAIT,'git-syncrepo.sh',['git-syncrepo.sh',dest_path])
        if retv:
            log('git','Error:update-failed',url)
            print "Failed to sync remote:",dest_path,retv
            return False
        log('git','updated',url)
        return True
    else:
        log('git','in-cache',url)
        print "[cloneserver.py:] Using cached git repo."
    return True

def handle_req(sock):
    while 1:
        line = read_line(sock)
        if line.endswith("\012"):
            line = line[:-1]
        if line.endswith("\015"):
            line = line[:-1]
        if line.startswith('wget '):
            if logged('wget',wget_clone,line[5:]):
                sock.send("ok "+line[5:]+'\015')
            else:
                sock.send('fail\012')
            sock.close()
            return
        elif line.startswith('git '):
            if logged('git',git_clone,line[4:]):
                sock.send("ok "+line[4:]+'\015')
            else:
                sock.send('fail\012')
            sock.close()
            return
        elif line.startswith('svn '):
            if logged('svn',svn_clone,line[4:]):
                sock.send("ok "+line[4:]+'\015')
            else:
                sock.send('fail\012')
            sock.close()
            return
        else:
            sock.send('fail\012')
            sock.close()
            return

peer = sock.accept()
while peer:
    print "Connected",peer[1]
    URL = 'nil'
    HOST = 'nil'
    handle_req(peer[0])
    peer = sock.accept()
